### ISSUE TITLE
undefined reference to `METIS_NodeND'

### ISSUE THREAD

**Issue by jiapei-nexera** [ceres-solver/ceres-solver#827]
Please refer to [https://github.com/KarypisLab/METIS/issues/33](https://github.com/KarypisLab/METIS/issues/33)

Trying to build [https://github.com/ceres-solver/ceres-solver](https://github.com/ceres-solver/ceres-solver) without **SuiteSparse BUT METIS** , I got: 

```console
undefined reference to `METIS_NodeND'
```

**Comment by sandwichmaker** [comment]
@jiapei-nexera we are working on fixing our build issues. cc: @sergiud

**Comment by jiapei-nexera** [comment]
@sandwichmaker 

By the way, did you notice that [https://github.com/DrTimothyAldenDavis/SuiteSparse](https://github.com/DrTimothyAldenDavis/SuiteSparse) also depends on **metis-5.1.0** ??

**Comment by sandwichmaker** [comment]
Yes i am aware of that.

**Comment by jiapei-nexera** [comment]
@sandwichmaker 
Could it be possible it's an issue from **METIS itself**? 
Take a look at [https://github.com/mfem/mfem/issues/698](https://github.com/mfem/mfem/issues/698) ?

**Comment by jiapei-nexera** [comment]
@sandwichmaker 

And also, current **METIS** is now split into two parts:

- [https://github.com/KarypisLab/METIS](https://github.com/KarypisLab/METIS) itself
- [https://github.com/KarypisLab/GKlib](https://github.com/KarypisLab/GKlib)

For now, I even failed to build **CERES examples**, as well as **dense_linear_solver_benchmark** with the following error message (but everything else built successfully. )

```console
[ 30%] Linking CXX executable ../bin/helloworld_numeric_diff
cd ....../ceres-solver-2.1.0/build/examples && /usr/bin/cmake -E cmake_link_script CMakeFiles/helloworld_numeric_diff.dir/link.txt --verbose=1
collect2: error: ld returned 1 exit status
make[2]: *** [examples/CMakeFiles/helloworld.dir/build.make:103: bin/helloworld] Error 1
make[2]: Leaving directory '....../ceres-solver-2.1.0/build'
make[1]: *** [CMakeFiles/Makefile2:6234: examples/CMakeFiles/helloworld.dir/all] Error 2
/usr/bin/c++   -Wmissing-declarations -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -O3 -DNDEBUG CMakeFiles/helloworld_numeric_diff.dir/helloworld_numeric_diff.cc.o -o ../bin/helloworld_numeric_diff  -Wl,-rpath,....../ceres-solver-2.1.0/build/lib:/usr/local/lib ../lib/libceres.so.2.1.0 /usr/local/lib/libglog.so.0.6.0 /usr/lib/x86_64-linux-gnu/libgflags.so.2.2.2 -lpthread 
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_randint32'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcorePop'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_rmpath'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_free'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreCreate'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_CPUSeconds'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_jbufs'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_siguntrap'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_idxsmalloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_cur_jbufs'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_errexit'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc_init'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreDestroy'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_randinit'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_realloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_log2'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcorePush'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc_cleanup'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreMalloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_sigtrap'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/CMakeFiles/rosenbrock.dir/build.make:103: bin/rosenbrock] Error 1
make[2]: Leaving directory '....../ceres-solver-2.1.0/build'
make[1]: *** [CMakeFiles/Makefile2:6338: examples/CMakeFiles/rosenbrock.dir/all] Error 2
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_randint32'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcorePop'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_rmpath'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_free'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreCreate'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_CPUSeconds'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_jbufs'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_siguntrap'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_idxsmalloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_cur_jbufs'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_errexit'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc_init'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreDestroy'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_randinit'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_realloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_log2'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcorePush'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc_cleanup'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreMalloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_sigtrap'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc'
collect2: error: ld returned 1 exit status
```

and

```console
[100%] Linking CXX executable ../../bin/dense_linear_solver_benchmark
cd ....../ceres-solver-2.1.0/build/internal/ceres && /usr/bin/cmake -E cmake_link_script CMakeFiles/dense_linear_solver_benchmark.dir/link.txt --verbose=1
/usr/bin/c++   -Wmissing-declarations -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -O3 -DNDEBUG CMakeFiles/dense_linear_solver_benchmark.dir/dense_linear_solver_benchmark.cc.o -o ../../bin/dense_linear_solver_benchmark  -Wl,-rpath,/usr/local/lib:/opt/SuiteSparse/lib:/opt/intel/oneapi/tbb/latest/lib/intel64/gcc4.8:/usr/local/cuda/lib64:/opt/intel/oneapi/mkl/latest/lib/intel64 /usr/lib/x86_64-linux-gnu/libbenchmark.so.1.6.1 ../../lib/libceres_static.a /usr/local/lib/libglog.so.0.6.0 /usr/lib/x86_64-linux-gnu/libgflags.so.2.2.2 -lpthread /opt/SuiteSparse/lib/libspqr.so /opt/SuiteSparse/lib/libcholmod.so /usr/local/lib/libmetis.so /opt/SuiteSparse/lib/libamd.so /opt/SuiteSparse/lib/libcamd.so /opt/SuiteSparse/lib/libccolamd.so /opt/SuiteSparse/lib/libcolamd.so /opt/SuiteSparse/lib/libsuitesparseconfig.so -lrt /opt/intel/oneapi/tbb/latest/lib/intel64/gcc4.8/libtbb.so /usr/local/cuda/lib64/libcudart_static.a -ldl /usr/lib/x86_64-linux-gnu/librt.a /usr/local/cuda/lib64/libcublas.so /usr/local/cuda/lib64/libcusolver.so /usr/local/cuda/lib64/libcusparse.so /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_intel_lp64.so /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_intel_thread.so /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_core.so /usr/lib/x86_64-linux-gnu/libiomp5.so -lm -ldl 
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_randint32'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcorePop'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_rmpath'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_free'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreCreate'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_CPUSeconds'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_jbufs'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_siguntrap'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_idxsmalloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_cur_jbufs'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_errexit'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc_init'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreDestroy'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_randinit'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_realloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_log2'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcorePush'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc_cleanup'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_mcoreMalloc'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_sigtrap'
/usr/bin/ld: /usr/local/lib/libmetis.so: undefined reference to `gk_malloc'
collect2: error: ld returned 1 exit status
```

**Comment by sergiud** [comment]
Ceres METIS linker error was fixed in 0c88301e6658f9c52e54621d45e333dce35c09ec. GKlib linker errors are unrelated to this issue.

**Comment by jiapei-nexera** [comment]
@sergiud 

Great... But **NOT** able to build with **benchmark**, which is installed from Ubuntu 22.04 repo directly.
And I believe you have noticed my other issue at: [https://github.com/sergiud/SuiteSparse/issues/5](https://github.com/sergiud/SuiteSparse/issues/5) ? The same as **ceres**.

May I ask where did you guys manage these packages?

```console
[ 99%] Linking CXX executable ../../bin/dense_linear_solver_benchmark
cd ....../ceres-solver/build/internal/ceres && /usr/bin/cmake -E cmake_link_script CMakeFiles/dense_linear_solver_benchmark.dir/link.txt --verbose=1
/usr/bin/c++   -Wmissing-declarations -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -O3 -DNDEBUG CMakeFiles/dense_linear_solver_benchmark.dir/dense_linear_solver_benchmark.cc.o -o ../../bin/dense_linear_solver_benchmark  -Wl,-rpath,/usr/local/lib/SuiteSparse:/usr/local/cuda/lib64 /usr/lib/x86_64-linux-gnu/libbenchmark.so.1.6.1 ../../lib/libceres_static.a /usr/lib/x86_64-linux-gnu/libglog.so.0.4.0 /usr/lib/x86_64-linux-gnu/libunwind.so /usr/lib/x86_64-linux-gnu/libgflags.so.2.2.2 -lpthread /usr/local/lib/SuiteSparse/libspqr.so.2.0.9 /usr/local/lib/SuiteSparse/libcholmod.so.3.0.14 /usr/local/cuda/lib64/libcublas.so /usr/local/lib/SuiteSparse/libsuitesparseconfig.so.5.12.0 -lm /usr/local/cuda/lib64/libcudart_static.a -ldl /usr/lib/x86_64-linux-gnu/librt.a /usr/local/cuda/lib64/libcublas.so /usr/local/cuda/lib64/libcusolver.so /usr/local/cuda/lib64/libcusparse.so /usr/lib/x86_64-linux-gnu/liblapack.so /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/libf77blas.so /usr/lib/x86_64-linux-gnu/libatlas.so -Wl,-rpath-link,/usr/local/lib/SuiteSparse 
/usr/bin/ld: ../../lib/libceres_static.a(eigensparse.cc.o): undefined reference to symbol 'METIS_NodeND'
/usr/bin/ld: /usr/local/lib/libmetis.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [internal/ceres/CMakeFiles/dense_linear_solver_benchmark.dir/build.make:118: bin/dense_linear_solver_benchmark] Error 1
make[2]: Leaving directory '....../ceres-solver/build'
make[1]: *** [CMakeFiles/Makefile2:369: internal/ceres/CMakeFiles/dense_linear_solver_benchmark.dir/all] Error 2
make[1]: Leaving directory '....../ceres-solver/build'
make: *** [Makefile:149: all] Error 2
```

**Comment by sandwichmaker** [comment]
is this issue still open?

**Comment by sergiud** [comment]
No longer an issue. The linker error was fixed in #808.
